cryptographic_estimators/UOVEstimator/UOVAlgorithms/collision_attack.py

```diff
@@ -0,0 +1,170 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+from ..uov_algorithm import UOVAlgorithm
+from ..uov_problem import UOVProblem
+from ...MQEstimator.mq_estimator import MQEstimator
+from ...base_algorithm import optimal_parameter
+from ...helper import round_or_truncate
+from ...base_constants import BASE_MEMORY_BOUND, BASE_NSOLUTIONS, BASE_BIT_COMPLEXITIES, BASE_EXCLUDED_ALGORITHMS
+from math import log2, e
+
+
+class CollisionAttack(UOVAlgorithm):
+    """
+    Construct an instance of CollisionAttack estimator
+
+    Collision attack is a general attack which works against any signature which 
+    follows the hash-and-sign paradigm. 
+
+    INPUT:
+
+    - ``problem`` -- an instance of the UOVProblem class
+    - ``gray_code_eval_cost`` -- logarithm of the cost to evaluate one polynomial in one vector using Gray-code enumeration (default: log(n))
+    - ``X`` -- Number of preimages
+    - ``Y`` -- Number of variables in the salt space
+    - ``memory_access`` -- specifies the memory access cost model (default: 0, choices: 0 - constant, 1 - logarithmic, 2 - square-root, 3 - cube-root or deploy custom function which takes as input the logarithm of the total memory usage)
+    - ``complexity_type`` -- complexity type to consider (0: estimate, 1: tilde O complexity, default: 0)
+
+    """
+
+    def __init__(self, problem: UOVProblem, **kwargs):
+        super().__init__(problem, **kwargs)
+
+        self._name = "CollisionAttack"
+        self._attack_type = "forgery"
+        self._alpha = 1.25
+        q = problem.order_of_the_field()
+        self._gray_code_eval_cost = kwargs.get("gray_code_eval_cost", log2(q))
+
+
+    @optimal_parameter
+    def X(self):
+        """
+        Return the optimal `X`, i.e. no. of inputs (preimages)
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.UOVEstimator.UOVAlgorithms.collision_attack import CollisionAttack
+            sage: from cryptographic_estimators.UOVEstimator.uov_problem import UOVProblem
+            sage: E = CollisionAttack(UOVProblem(n=24, m=10, q=2))
+            sage: E.X()
+            11.95
+
+        """
+        _, m, q = self.problem.get_parameters()
+        alpha = self._alpha
+        r = self._gray_code_eval_cost
+        X = 9.25  +  (1 / 2) * (-log2(3 * m * r)  + log2(alpha) + m * log2(q))
+        precision = 2
```

Comment:
I dont know where those constants are taken from, but maybe its worth to hardcode them in the constructor as class/static variables? Specially `precision` looks weird to me. Why is not the python internal `round(X, 2)` used? Why round at all?

Reply:
Changed to `round(X, 3)`